### ISSUE TITLE
World of Warcraft - Test results

### ISSUE THREAD

**Issue by IngeniousDox** [acomminos/wine-pba#7]
**commit:** https://github.com/acomminos/wine-pba/commit/0ca7edba117f5c91c75623880caf8e4d1eeb117b0ca7edba117f5c91c75623880caf8e4d1eeb117b
**specs:** 
- CPU: Intel i5-4670K OCed @ 4Gz
- GPU: NVidia 970GTX OCed @ 1565Mhz
- RAM: 16gb @ 2400Mhz

**Testing Spot:** Suramar, on top of Astravar Harbor
**Test Reults:**

```
Staging 2.21 DX11: 42 fps (no csmt - lowers fps)
Staging 3.2 DX11: 18 fps (no csmt - lowers fps)
wine-pba 2.21 DX11: 67 fps 

Staging 2.21 DX9: 48 fps (no csmt - lowers fps)
Staging 2.0 DX9: 82 fps
wine-pba 2.21 DX9: 97 fps
```
Amazing results for DX11!! DX9 is also a bit faster then before (7 fps higher then your first build). But the improvement is DX11 is amazing. Eyeballing the CPU% from HTOP and the GPU-Util from nvidia-smi, I get:

```
DX9: 275% CPU / 67% GPU for 97 fps
DX11: 256% CPU / 51% GPU for 67 fps
```
Way higher then before. Looking forward to your upcoming improvements.

Dox

**Comment by IngeniousDox** [comment]
**BUG:** I think you have a memory leak or similar. My system locked up. After restarting I kept an eye on MEM% using HTOP, and WoW is increasing with 0.1% every 2 seconds.

**Comment by IngeniousDox** [comment]
I have also noticed something different, the fps I get after I log in from a fresh start, on my test toon varies. I usually have around 65 fps, but I have also had 70+ fps a few times, but I have also noticed 45ish fps a few time. This interested me.

I start the game offscreen on a different tag of my AwesomeWM setup. My background fps is capped at 30, but my foreground fps isn't. I normally have VSync on with triplebuffer, but for testing I disable vsync. I figure it has something to do with how fast I switch to the game or something, and the average fps it has been running on till I fully loaded in the game. So I tried without background limiting, that seems to get me to around 60 fps (Which is probably because for offscreen rendering it is vsynced automagically to my desktop refresh rate by either my WM or Xorg or Nvidia, so the average became 60 fps).

So then I disabled fps limiting altogether, made sure I was on that Tag from the moment the game starts up. And low and behold, I get 75 fps in that spot consistently. This was remarkable enough, that I figured I would post here.

Dox

**Comment by ikidd** [comment]
Very happy with this on older i5 laptop with nVidia Optimus.  Went from 8fps to over 30 in Dalaran, from 30 fps in world PVE to 60fps.  Made WoW useable for me without going to Windows boot on laptop.  But huge memory leak makes system grind to a halt within a few minutes.

Doesn't work at all on AMD gpu system.

**Comment by acomminos** [comment]
Will address the memory leak issue asap in #8.

**Comment by acomminos** [comment]
Leak should be fixed now.

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/a5348cf24854102f49ba6e227e6e837a3871fa35

**result on regular testing spot:** 80 fps / 260% CPU / 62% GPU

Another increase in fps. The leak seems fixed.

Thanks for the quick fix.

**Comment by yaomtc** [comment]
It's the opposite result for me on AMD. Using a Radeon RX 580 with an Intel i5 4690k, with Linux 4.15.6, Mesa 17.3.6, and the pba-2.21 build from Lutris 0.4.14. When compared to unpatched Wine Staging 2.21 from Lutris (~70-80 fps) with the same static camera angle and position, there is a major decrease in frame rate, down to around 20-26 fps. All options in Lutris are set to the default, with the exception of disabling Lutris runtime in each test.

**Comment by acomminos** [comment]
Going to look into using `AMD_pinned_memory` where available- I've heard it's much faster than `ARB_buffer_storage` on AMD cards.

@IngeniousDox, I've made a variety of changes to the data structures backing the buffer heap that should make it much more efficient- can you measure the impact on your benchmark when you have a chance?

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/57f67be9a6e5af59039321cf7ddf55bbad838e16

Unfortunatly I moved my toon slighty. So first a new baseline test with the previous commit:

base: 74 fps / 255% CPU (1 core maxed) / 57% GPU

I'll find a better spot for testing, or /timetest a route in the future.

test result: 72ish or 64ish fps / 244% CPU / 56% or 44% GPU
test result: 69ish or 62ish fps / 244% CPU / 56% or 44% GPU

FPS is slightly lower, and it still depends on how fast I log in I think. And there was a weird jumping up and down every few seconds, my highest CPU core % during that time dropped from 95% to below 80% for a bit.

Yes, need a better way to test this.

**Comment by IngeniousDox** [comment]
Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Isles):
```
Base -> Newest
Min: 20 -> 14
Max: 122 -> 113
Average: 72 -> 64
```
And back:
```
Base -> Newest
Min: 18 -> 15
Max 149 -> 142
Average: 76 -> 69
```
Had to go back and forth 2x, since the nvidia shadercache is being filled the first time I go. Does show fps is down a bit. But before we can really test, it would be lovely if the fps you get when you log in, doesn't depend on how fast I log in, or if I have it focussed. ;)

**Comment by IngeniousDox** [comment]
Did a quick lfr, killing the game midfights to switch between the 2 pervious build and the current build. I had the gut feeling the previous build I had is slighty faster then the new build, but it hardly matters in raids. FPS is lower the out in the world, GPU has lower util, while CPU has more usage.

I just want to note, that I have double the fps compared to Staging 2.21 without PBA in raids now aswell. So thats great.

**Comment by acomminos** [comment]
Thanks for the data- I'm going to perform further analysis of the addresses in fenced batches. I suspect coalescing batches prior to submission into the primary free tree should give us the best of both worlds.

**Comment by acomminos** [comment]
I've implemented deferred coalescing in ac9b10e, which should vastly improve games (such as WoW) with cyclic buffer allocation patterns.

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/68de8e9b3f26e68bc6d64f353e0954ddab2f7590

First noting, that logging in still ends up giving me either around 60ish fps or  around 70ish. When I have 60 fps, logging to another toon, moving around a bit, and then log back, and it is still on 60 fps. Both have around the same CPU usage, but the 60 fps has 44% gpu while 70 fps has 54% gpu.

`test result: 71 fps / 267% CPU / 52% cpu`

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore):
```
Min: 20 fps
Max: 106 fps
Average: 67 fps
```
And back:
```
Min: 18 fps
Max: 136 fps
Average: 71 fps
```
Now on a more important note: Game feels smooth(er). First without vsync, then with vsync back on. It isn't something you can put fps numbers on, but it is a gamer feeling I guess.

**Comment by acomminos** [comment]
Thanks. The phenomena you're noticing is likely to be an improvement in consistency of frame timing. This makes sense, since we no longer freeing large chunks of buffers periodically (deferred instead).

I'll look into a good way to quantify this- I believe `glxosd` can report this.

**Comment by IngeniousDox** [comment]
Quick 2 boss lfr Aggr/Argus: Feels smooth, even under load of 25 man raid. It was already feeling smooth on the "https://github.com/acomminos/wine-pba/commit/a5348cf24854102f49ba6e227e6e837a3871fa35" build that I was using yesterday (The build that gave me the best fps thusfar). But this feels better.

I think it will really benifit classes that stuff like blink / charge / fel rush etc. Heroic leaping on warrior feels much better.

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/c565623357098b1c946b4ef6c7768f123630e27f
```
test result: 71-75ish fps / 280-290%CPU (1 core 95%) / 57%-59% GPU
```
I guess I was "lucky" with timing this log in, and getting high fps. I tested 2 more times, and got slightly lower fps ranges. It remains an interesting bug.

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore):
```
Min: 22 fps
Max: 106 fps
Average: 66 fps
```
And back:
```
Min: 22 fps
Max: 137 fps
Average: 71 fps
```
FPS on timetest didn't change significantly.

**Comment by IngeniousDox** [comment]
PS: It feels "slightly" less smooth to turn around with this build. I had to relog both builds a few times to double check. With the previous build Heroic Leaping and turning around to charge back to a dummy also feels better.

What `glxosd` output are you looking for to make testable?

**Comment by IngeniousDox** [comment]
In Warrior class hall: Charge dummy, heroic leap away, turn 180, charge back, mortal strike.

old = previous build (https://github.com/acomminos/wine-pba/commit/68de8e9b3f26e68bc6d64f353e0954ddab2f7590)
new = latest build (https://github.com/acomminos/wine-pba/commit/c565623357098b1c946b4ef6c7768f123630e27f)

(once with vsync on, once with vsync off)

[glxosd_benchmarks.tar.gz](https://github.com/acomminos/wine-pba/files/1774610/glxosd_benchmarks.tar.gz)

I had between 120 and 130 fps without vsync.

While I was typing this comment I had the game running, and I tabbed back and saw 150 fps. There were no other warriors anymore around me anymore, so I quickly redid the tests without Vsync:

[glxosd_benchmarks_solo.tar.gz](https://github.com/acomminos/wine-pba/files/1774654/glxosd_benchmarks_solo.tar.gz)

I know you have WoW, and you can get this glxosd TimeRecorder output yourself while knowing exactly what you recorded. But since I wanted to toy with it a bit, and got some output, I figured I would just give you my datasets anyways.

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/a5dff624a8a63dca94138145105718ee712f7ddf

Test results similar. Not going repost. However, the login lower fps issue seems fixed.

Closed: https://github.com/acomminos/wine-pba/issues/19

**Comment by acomminos** [comment]
I'm interested to see how 0e1e946 performs for you; I noticed that UBO updates were bottlenecking PBA-allocated buffers, so I integrated `ARB_multi_bind` to reduce the number of driver calls needed for state updates.

It seems to be working very well for me (qualitatively).

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/0e1e946b099e15b09430be8f2334b6d37e46b311 (Well, the next commit where you changed README)

```
test result: 75ish fps / 280% cpu / 54% GPU
``
CPU usage distribution over my 4 cores is good. I don't have 1 core 100%.

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore):
```
Min: 19 fps
Max: 106 fps
Average: 77 fps
```
And back:
```
Min: 21 fps
Max: 134 fps
Average: 80 fps
```
Average is 10 fps higher, so really nice increase in fps.

Now for the holy crap moment: In skyhold @ the mission board, I usually was around 60/70 fps. Now I'm suddenly at 140/150. So yeah, FPS is up. I'll test real raiding this weekend, but I'll see if I can do a quick LFR somewhere today.

**Comment by IngeniousDox** [comment]
Quick LFR Aggr / Argus: Aggr felt about the same, Argus felt a bit lower.

Honostly, not really a good test, I'll see if I can test 2.21bpa vs 3.3pba on friday when I'm doing mythic progression.

**Comment by asharin75** [comment]
Might be a bit of both, did notice on wine-pba's github that they'd made some other changes too...As long as it works!

**Comment by xpander69** [comment]
Just came here to say Thank you for the awesome work. Wow is quite a bit more enjoyable now.

did some testing also:

https://www.youtube.com/watch?v=t7t61CScbGM


do not pay attention to the bad drops for 3.3 with pba, seems it was still filling the shader cache or something. im currently using 3.3 and its as smooth as the 2.21 was and i think it has even improved framerate a bit. might be placebo though :)

**Comment by IngeniousDox** [comment]
I saw you video. I figured it was filling shadercache as well. Anyways, FPS while doing /timetest is definitely higher. It definately isn't placebo.

What we need is now is test this in 20 man (mythic) raiding. To see how well it renders with a lot of spell effects. This is where the biggest hits in FPS have always been.

**Comment by IngeniousDox** [comment]
Switched from 4.15 Ubuntu Mainline kernel to Liquorix 4.15 kernel after someone poked me about having good results with it. Here is the timetest with Staging 3.3+PBA:

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore):
```
Min: 20 fps
Max: 112 fps
Average: 86 fps
```
And back:
```
Min: 21 fps
Max: 150 fps
Average: 92 fps
```
So, another 10ish fps higher again. Hurray! 

I quickly retested with Staging 2.21+PBA. That however didn't give the same fps increase, it was in the range of 1~2 fps more.

**Comment by Dehir** [comment]
With wine-3.3 + amd fix seems to improve my performance.

https://www.youtube.com/watch?v=FwAo6CsckCA 

LFR raiding 25+ ppl get around 30~ fps.

**Comment by IngeniousDox** [comment]
Varimatras 20 man mythic: Staging 2.21 gets slightly more fps (35ish) then Staging 3.3 (32ish) under raid load (similar results as with the other games). But staging 3.3 simply feels better / smoother. Staging 2.21 has some dips, and just general moving with 3.3 just feels better during raids.

Now, could have to do with the fact that 3.3 has newer code....but w/e the reason, I'm going to stick with 3.3 + PBA for WoW.

**Comment by ikidd** [comment]
Huh, thought I'd posted this but I don't see it in the thread.

Flightpath Crimson Thicket to Vengeance Bay.  I ran it a few times to get the shaders settled, then timed.  DX11 only.

Laptop:  nVidia 630M on Optimus, i5-3910 with 8GB ram
                  slow/fast/avg
3.3PBA:         4/145/23
3.3Stg:           2/45/20
3.3:             Fails to load game
2.21PBA:       3/176/46
2.21Stg:         2/58/30

Desktop:  AMD 7770, Pentium G6950, 8GB ram

3.3PBA:        3/180/40
3.3Stg:          2/70/20
3.3:           Fails to load game
2.21PBA:  Unusable
2.21Stg:        1/37/8

So AMD is great on 3.3PBA where it didn't work at all on 2.21PBA, nVidia laptop best on 2.21PBA by large margin.  I ran the nVidia one several times because I wasn't trusting the result, but that's the consistent outcome.

**Comment by acomminos** [comment]
Thanks for the data- I wonder how much (if any) slowdown is being introduced by non-pba-related 3.3 changes.

I'm going to rebase the latest patches onto 2.21 tonight to figure out what's up.

**Comment by acomminos** [comment]
1631e46 shouldn't improve benchmark numbers significantly, but it should reduce stuttering.

**Comment by IngeniousDox** [comment]
Honostly, I think the we are working the margins of CPU bound / GPU bound now. I'm thinking 3.3 uses more CPU then 2.21. So if you are CPU bound you get slightly less fps. Like during raiding.

During /timetesting, I think you run against the max of your GPU now. And well, that might be just my setup, and could be different for other setups. I definitely got increased fps with 3.3 during timetesting. Especially after I switched to zen kernel for throughput, but that is CPU throughput allowing my GPU to do more work. I still need to test raiding without zen kernel, since throughput sacrifices fps a bit, and I get CPU bound.

So I'm compiling your latest patches where you said you are using less CPU. And I'll timetest soonish. And I'll test raids tonight.

**Comment by xpander69** [comment]
funny thing is that i get more FPS with 3.3 on heavy cpu situations, like Dalaran with lots of people i get 40+ FPS now, while with 2.21 i had drops down to 25 at times, Rest of the FPS seems pretty much same. I haven't done any timetests though. Just compiled latest patches also will see whats it now

**Comment by IngeniousDox** [comment]
**commit:** https://github.com/acomminos/wine-pba/commit/4b64220635c9eb0aeee34c451421085d1a71bb6b

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore):
```
Min: 20 fps
Max: 110 fps
Average: 83 fps
```
And back:
```
Min: 22 fps
Max: 139 fps
Average: 87 fps
```
I actually lost fps compared to my previous test with Zen kernel. But I'm not worried about that. I'll test it in raiding tonight for a proper test.

**Comment by Dehir** [comment]
@IngeniousDox My guess is the same even havent tested with 2.21 thou. 

Ill get more fluid performance even high crowd raiding when fps dips close to 20~35. There is no stuttering and spikes any what so ever. GPU usage actually increases. Noted that running my ryzen 7 1700 with 3.0Ghz stock vs 3.9 overclock makes difference. It actually affects many game framerates. Chivalry Mediaval Warfare example.

But general for me i think its playable at the current state even while raiding with amd 7 1700 + RX 480. Thou there are always for improvements, and you newer know if there becomes patch to break things on blizzard side ;(

**Comment by xpander69** [comment]
the new patchset seems to be a miracle in heavily crowded places like dalaran 50+ FPS and Orgrimmar in the middle near bank 60+ FPS, was about 10-20FPS worse before, will keep tracking this but it seems really good so far . My Ryzen 7 1700X is cloced at 3.9ghz, XFR and boosts are disabled, performance governor is enabled when gaming.

**Comment by IngeniousDox** [comment]
I'm confirming increased FPS in Dalaran. Quickly set up a 2nd prefix so I could switch fast. Found a spot in Dala where I had 50 fps, relogged, and ended up ~10 fps higher there. Going to find a LFR during lunch.

@Dehir Overclocking CPU helps indeed. I have 4 cores (No HT). WoW is limited by the singlecore mainthread. CSMT gives it a 2nd thread/core to do the GL calls, and NVidia Threaded gives the driver another thread/core to do CPU intensive stuff. So thats 3 cores. The 4th core does sound / OS for me I guess.

You could almost say I bought my CPU based on overclocking especially for WoW on linux with CSMT + Nvidia threaded....and you are correct, that is actually what I did 3 or 4 years ago, and it is still working great. ;)

**Comment by xpander69** [comment]
you use GL_THREADED on top of CSMT? that helps even more? .,..hmm never thought about this. i have 16 threads and i see them mostly balanced in load, wow is totalling 12-14% on those from 100%(all 16 threads) when in cpu limited situation thoug then i can see 1 core going up to 100%

**Comment by IngeniousDox** [comment]
They are 2 different technologies, like I described. It depends on the game if they work together though. Only way to find out is to test. With Staging 2.0 and DX9 it worked fine...but later versions CSMT just added CPU usage so I stopped using CSMT, and only use GL_THREADED.

With PBA CSMT works fine again, and GL_THREADED still works fine.

Now about raiding, just did LFR, and on Varimathras I had 40+ fps, 10 fps more then on Wed night. So yeah deffo better.

**Comment by IngeniousDox** [comment]
Ok, going to have to rethink something in testing. I figured I would double check if GL_THREADED actually gives me more fps. Since right now CSMT is working great, it could very well be that I'm just wasting CPU time by using GL_THREADED aswell. And I got some interesting results.

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore):
```
Min: 20 fps
Max: 116 fps
Average: 88 fps
```
And back:
```
Min: 22 fps
Max: 160 fps
Average: 98 fps
```
This is actually the highest I have ever had...but I figured, it could be a fluke, so I'm going to recheck it. I went back and forth mutliple times and every time I did this, I actually got lower averages:
```
Forward: 88.2 > 86.7 > 84.6 > 83.2 > 83.2 > 84.0
Backward: 98.4 > 96.6 > 95.0 > 91.6 > 90.9 > 92.0
```
Restarted the game to see if I would till get it, but this time I stayed on Forward ~88, Backwards ~98 for 5 times back and forth. So it seems I can't repeat it. But I figured I would note it.

Tested 1 more time by pressing Alt-Z right after I click the other FP, this removes the UI. This gave me Forward 94, Backwards 104. Perhaps I should have always tested like this, but I did't.

Anyways, my conclusion is that I (and you) need to check with and without  __GL_THREADED_OPTIMIZATIONS. It has always given me more fps in the past years, but this seems no longer the case.

**Comment by IngeniousDox** [comment]
Ok, I confirmed I'm not crazy. I used the previous build to test the difference with and without __GL_THREADED_OPTIMZATIONS.
```
GL_THREADED: Forward 83, backwards 87.
No GL_THREADED: Forward 81, backwards 91
```
So on the previous build it actually smooths out the fps average. 

This also confirms, that [4b64220](https://github.com/acomminos/wine-pba/commit/4b64220635c9eb0aeee34c451421085d1a71bb6b) is an amazing improvement.

**Comment by SteveEbey73742** [comment]
Flight path tests using vulkan as D3D render in place of opengl in wine registry.
Dalaran (BS) - shipwreck cove 8 min 186 max 58 avg
Booty Bay - Lights hope 7 min 201 max 88 avg
Moonglade - Dig site, Tanaris 11 min 195 max 82 avg
Wildhammer Stronghold (outland) - Stormspire, Netherstorm 9 min 149 max 54 avg
Valiance Keep - Westfall Brigade (Northrend) 6 min 143 max 64 avg
Timeless isle - Shado Pan Garrison 1 min 265 max 54 avg
Srotmshield - Rizlits Hold Fast, Nagrand (draenor) 7 min 103 max 55 avg

Compiled 64 bit only, CFLAGS=-march=bdver2 -mtune=bdver2 -O3 -fPIC -fomit-frame-pointer -fno-align-functions -fno-align-loops -pipe
symlinked wine to wine64, run regedit, put vulkan in place of opengl in Direct3d render. Enabled CSMT on the staging tab of winecfg.
Launch wow from desktop shortcut, not battle.net app.
Running Geforce GTX 970, 4Gig DDR5, 390.25 proprietary Nvidia from website. AMD FX 6350 6 core, 3.6Ghz, 16Gig DDR3 1600 ram. WoW set on 7 for all.

**Comment by IngeniousDox** [comment]
It would be lovely if it was that easy, and we could put Vulkan as renderer. DXVK is being made for that reason, but doesn't work for WoW yet.

You tested with OpenGL and Wine-pba.

**Comment by SteveEbey73742** [comment]
No, I did not.  What do I need to show, I have screen shots of both, and can see visual differences in running vulkan versus opengl.  is there a wine test somewhere, that can show the underlying render that is being used.

**Comment by IngeniousDox** [comment]
Well, if you found a way to already run WoW on Vulkan, people will be interested. So we will want to know the steps you took, so we are able to replicate it. I'll start with a few questions, perhaps you can fill in the blanks:

- What did you compile? I'm assuming Staging 3.3 + wine-pba
- What vulkan did you use? Wine 3.3 has stubbed out Vulkan support, did you add Vulkan in another way?
- What registery key did you change?
- If you used DX11, did you use DXVK? If not, what other project is there that translates d3d DX11 to Vulkan?
- If you used DX9, did you use VK9? If not, is there another project that translates d3d DX9 to Vulkan?
- And sure, I would like to see the screenshots. Can post them here.

**Comment by TomBZombie** [comment]
wine-pba wouldn't work with vulkan anyway,  it makes use of opengl features.

**Comment by mirh** [comment]
Can we please keep OTs out?
People is not here to hear about vulkan.

**Comment by IngeniousDox** [comment]
Its a thread about World of Warcraft - Test results. Its on topic enough. We are testing WoW in various ways, with and without PBA. Now, I haven't heard of a way to get WoW working it with Vulkan, so I'm open to hearing about it. And if I works with Vulkan, I'm going to test with it to get comparisons. You are free to mute the thread if you ain't interested.

**Comment by mirh** [comment]
> with and without PBA

And that's it. 
If you really want to sport "best case" results, reboot in Windows and check it. 

If you care about vulkan (which again, has nothing to do with this repo), discuss that on https://github.com/doitsujin/dxvk/issues/14

**Comment by TomBZombie** [comment]
Assuming dxvk works with WoW (last I heard they hadn't implemented dx10 so it doesn't) it would still be interesting to compare dxvk vs wine-pba performance. I'm sure nobody here really cares whether wine is using vulkan or opengl for its renderer, we just want the best fps we can get.

**Comment by mirh** [comment]
Yes, sure, but even assuming DXVK was perfect, even though it uses the noblest® of the APIs, I don't think it could surpass *native* dx10/11 windows drivers (which I'll just want to remember everyone: have undergone decades of optimizations - more often than not, even with game-specific hacks)

In this sense, once you benchmarked X scene of the Y game on Z hardware once, you are done with your "setting a goal" aim. 
(and tbh, it's really sad nobody hasn't still reported such results)

**Comment by yaomtc** [comment]
Took me a while to get around to editing this, but I tried a more recent build (34a2a05 I think) and got some decent frame rate _improvements_ this time, compared to Staging. Here you go.

https://youtu.be/sYMb4ZXqSbI

**Comment by SveSop** [comment]
The benchies i have done with Unigine benchmarks comparing DXVK vs. wine-pba, shows wine-pba slightly ahead. (API set to DX11 in Unigine Valley and Heaven).

Using wine-staging-pba + dxvk makes no difference it seems, as "pba" patches wined3d, and afaik dxvk kinda skips using functions from wined3d?

@SteveEbey73742 
You can post the output from vulkaninfo when running wine (from the VulkanSDK) to "prove" you are able to run vulkan apps under wine-staging-pba-3.3.

Its not enough to just change a registry setting to get DirectX to render over Vulkan instead of OpenGL.

**Comment by ardacam** [comment]
**Specs**

- **CPU:** AMD FX-8350@4.0GHZ
- **GPU:** AMD R9 380 2GB
- **OS:** Solus OS Budgie
- **RAM:** 8GB
- **Graphics Preset:** 7
- **AddOns:** 
AdvancedInterfaceOptions
ArtifactPowerUser
AskMrRobot
BagBrother
Bagnon
BigDebuffs
BigWigs
flyPlateBuffs
LittleWigs
NiceDamage
nPlates
OmniBar
OmniCC
RSA
Scrap
SUI
WorldQuestTracker

I removed modules from list since they are situational.

I did greater invasion point, dungeons, world quests, random bgs and 2v2 arenas. Lowest fps was in **greater invasion point** which was around 15-20. **View Distance** seems to be the most effecting option in open world. Setting it to 4 and everything else on default gives 50+ fps in **Suramar**. Compared to **gallium-nine** (which is the superior choice for AMD GPUs most of the time) **wine-pba** is definetly better for WoW.

**Comment by Saroufim** [comment]
That makes me wonder, is it not possible to use pba and gallium nine
together?

**Comment by SveSop** [comment]
@Saroufim I suspect this is possibly the same reason why there is no change when using DXVK with or without "pba"... cos DXVK uses its own threading++.
Not looked at gallium-nine, but i guess it has its own wined3d/d3d9.dll++, that does this differently than regular wine? DXVK replaces dxgi.dll and d3d11.dll and uses its own vulkan render path, and is not "affected" by pba it seems.

I can do some more testing with DXVK and PBA when wine-dev gets all the vulkan patches implemented. Tested with staging-pba-2.21 and dxvk, but was no difference between that and regular wine-staging-2.21. As of late, pba is based on 3.3, and wine-vulkan-3.3 is using its own set of patches atm, so not too easy to test, although i guess it should not be too hard to implement pba patches on that sourcetree tho. (Might even patch straight up.. i dunno)

**Comment by Dehir** [comment]
Ill agree with @SveSop with this idea. DXVK and Gallium propably have their own code on direct3d and wine-pba doesnt directly affect those. But it would be great thing for both of those projects to take look at what wine-pba is trying to do and figure out would implementing wine-pba approac/patches improve their performance too if they are lacking these improvements?

**Comment by mirh** [comment]
> That makes me wonder, is it not possible to use pba and gallium nine together?

Not sure how much they "relate" to this, though they already have https://github.com/iXit/Mesa-3D/issues/297
And none of these patches have anything to do with threading, anyway. 

Also, for the love of god, instead of mentioning vk, why don't you benchmark windows already?

**Comment by SveSop** [comment]
@mirh 

> Also, for the love of god, instead of mentioning vk, why don't you benchmark windows already?

I plainly refuse to dual-boot windoze! :)

**Comment by acomminos** [comment]
I've made `GL_CLIENT_STORAGE_BIT` a mesa-only thing in 87307b1b9093c769d21d0803bb122b9866bb887c for now. Using DMA CACHED memory for buffers streaming on NVIDIA appears slower, so this should be a win for everyone.

**Comment by acomminos** [comment]
Also, given the earlier discussion, I should probably chime in and mention that DXVK will not benefit from any of my patches. DXVK implements D3D11 independent of wined3d (which is what the PBA patches apply to).

**Comment by mirh** [comment]
By the way, did you got to check whether AMD_pinned_memory is indeed faster than buffer_storage? 
I have seen on mesa-dev that it might even be supported by intel hardware.

**Comment by IngeniousDox** [comment]
**commit**: https://github.com/acomminos/wine-pba/commit/87307b1b9093c769d21d0803bb122b9866bb887c

I don't think I had issues with GL_CLIENT_STORAGE_BIT. At least not during timetest.

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore) - Min/Max/Avg fps:
```
CSMT+NV: 11 / 115 / 86
CSMT: 18 / 121 / 89
Prev Build: 18 / 120 / 88
```
And back:
```
CSMT+NV: 14 / 142 / 87
CSMT: 18 / 156 / 97
Prev Build: 18 / 162 / 97
```
I did each time 3 times, first to fill shadercache 2 and 3 to confirm I get the same results.

However, I have to note from eyeballing "watch nvidia-smi", I'm reaching 100% gpu usage from time to time during timetest. Which means I'm testing GPU bound there. Seems ok to compare builds I guess, but not really representative for actually playing the game.

Raiding is CPU bound, I have around 50% GPU usage then. And there fps is more important. I'll test mythic raiding tonight.

**Comment by SveSop** [comment]
@IngeniousDox So, with the latest commit, you have to turn off __GL_THREADED_OPTIMIZATIONS to get "back" to the performance you had before?

**Comment by IngeniousDox** [comment]
Not just the last, commit, it was already before that. __GL_THREADED doesn't help, it just adds CPU overhead you can't use anymore.

**Comment by SveSop** [comment]
@IngeniousDox 
I'm not seeing the same degradation with __GL_THREADED as you do in wow (or Unigine Benchmarks for that matter).
Tested the same flightpath you described over, zoomed character all in (1st person), and pressed ctrl+z to turn off UI elements.

Results:
```
Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore) - Min/Max/Avg fps:
NV=1    28.2/153.6/93.6
NV=0    35.9/133.7/63.9
Vengeance Point (Broken Shore) -> Crimson Thicket (Suramar) - Min/Max/Avg fps:
NV=1    32.9/192.7/98.9
NV=0    32.2/179.1/75.8
```
Not sure why you get worse performance with __GL_THREADED_OPTIMIZATIONS=1 tho.. Had the same decline in performance with Unigine Valley aswell.. Although i got a wee bit higher max fps in Unigine with NV=0, i lost average fps.

The only difference i can immediately think of (besides hardware like motherboard and gpu manufacturer) is that i use a different kernel perhaps? The distro SHOULD not have anything to say on this nVidia option, as i think we both use 390.25 driver...

**Comment by IngeniousDox** [comment]
Like I said, it is with my PC setup. I noticed I dropped in FPS, instead of gaining fps. So I tested. And in my case disabling __GL_THREADED gave me more fps during timetesting. I think it basically was due to my 4 CPU cores already being fully loaded. The extra CPU it takes to setup another thread for the driver, and the CPU time it takes to sync with that thread ended up costing me fps.

I encourage people to test for themselves.

I'm also going to restate that a /timetest isn't really a good test, but it is the best we have. Real playing happens in raids/bgs/arenas/dungeons. It might be that it helps me there. Atm I just don't play enough to really test it.

**Comment by SveSop** [comment]
I agree.. It is not that easy to come to a 100% conclusion from timetests alone.

I have done some testing with 4 vs 8 cores - ie. 4 "real" and 4 HT threads, and found that i got a bit better performance with 8 threads. I also run the wow binary with taskset -c 0-3, so that the 4 "real" cores gets the most of the load. There is something fishy going on with threads in wine vs. driver and whatnot, but doing that MIGHT leave the GL_THREADED option to offload on threads that are not in use by wine somehow.. i dont know.
Disabling HT in bios, and running only on the 4 real cores made for a bit worse performance, but then again, i did not test this with GL_THREADED 0 vs 1 at that time, and it might be something that effect this.

I did (not with pba tho) quite a few tests comparing kernel with PDS cpu scheduler vs. MuQSS, and found that MuQSS divides loads among threads more equal, but that is not really beneficial, as threads 0-3 is the "real" cores, and 4-7 is HT threads sharing off the real cores.. And i found that having cores 0-3 loaded 80+%, and 4-7 10-15% yielded more fps than all threads loaded 40-50%. I concluded then that PDS was more "sane" when distributing load amongst threads, as it would prioritize loading the actual real cores rather than treating all threads seemingly equal. This "function" is something related to "smt-nice", where the kernel is made aware of what cores is "real" vs. "ht", and it might just be that this is included in 4.15 kernel for all cpu schedulers.. dunno.

**Comment by IngeniousDox** [comment]
Ok, had some time to test raiding Coven Mythic tonight. Using __GL_THREADED_OPTIMIZATIONS gave me slight lower fps, mostly between 35 ~ 45 fps. Without that I ended up 40 to 55, and without it it just felt smoother.

Tested this build, and the previous 1. Didn't notice much difference once shadercaches were filled.

**Comment by SveSop** [comment]
A bit off-topic here, but since the wine-staging git repo does not have any "issues" ability.. how does one get in touch with the ppl doing the staging patches now really?
wine-staging-3.4 does not fix vsync, and it seems its back to wine-dev version of vsync - ie. its on no matter what you do, and locked at 60 fps (for a 60hz monitor i guess).

**Comment by mirh** [comment]
They.. [still seems](https://github.com/wine-staging/wine-staging) to recommend using the normal bug tracker?
In your case, I think [this](https://bugs.winehq.org/show_bug.cgi?id=44623) should be it.

**Comment by SveSop** [comment]
@mirh Hmm.. It does not seem as the patches from [https://bugs.winehq.org/show_bug.cgi?id=44623](https://bugs.winehq.org/show_bug.cgi?id=44623) has made it to submission.. I see Mr. Kucia posted this on the patches list to winehq.

Reporting vsync bug in staging on the bugtracker? I dunno... This vsync bug has been there for, well years, and wine-staging fixed this years ago. So its nice if this gets implemented some day, but for now, it does not work with staging either, so i guess ill test some more.

Might be me that is a dick, but "bugzilla" should honestly be studied by scientists.. especially now that Stephen Hawking is gone.. cos where else can you study a huge black hole up close..

**Comment by mirh** [comment]
Vsync bug is in wine first and foremost - and having bugs in a patch for a bug.. Still makes it related to the latter. 
Then it may have been there since ages, but for magic to happen somebody doing the right report with the right info is needed. 
(and I hope I don't have to remember you about what means "being unpaid and freelance")

Then it was questioned on the [mailing list](https://www.winehq.org/pipermail/wine-devel/2018-February/123436.html), and it might re-land next month, or tomorrow, who knows. 
If you want further info ping them directly on #winehq or #winehackers.

**Comment by SveSop** [comment]
Well.. functions not present in wine i think is not considered a "bug". Staging implemented a fix that probably did not pass some test = not implemented, and as you can see on that link you posted with comments on Kuzia's patch, they don't agree on the implementation. (Wine dev's that is).

Bug-reporting this on bugzilla when others already have AND wine-devs already have suggested patches = fruitless.

Suggesting i try to reach someone on IRC is the best bet, and should have been suggestion in your first post @mirh Thx.

**Comment by mirh** [comment]
> Well.. functions not present in wine i think is not considered a "bug"

Everything they are missing from actual windows is bug. 
And bumping threads with newer information is not pointless. 
...

Back on topic: has anybody even the slightest clue on whether current pba is good enough for perfection (aka performance parity) or not?

**Comment by SveSop** [comment]
Got a tip on #winehackers, and if anyone else is interested in forcing vsync off for 3.4, you can either edit dlls/wined3d/swapchain.c yourself and change:
```
-    swap_interval = swapchain->desc.swap_interval > 4 ? 1 : swapchain->desc.swap_interval;
+    swap_interval = 0;
```
Or, you can patch the attached patch after the pba patches :)

[0001-vsync-disable-hack.patch.gz](https://github.com/acomminos/wine-pba/files/1822170/0001-vsync-disable-hack.patch.gz)

**Comment by AndersDala** [comment]
Just curious, but is the patch in https://source.winehq.org/patches/data/142465 not also solving the same issue as above? (nevermind if it is "beautiful code" or not - ref to mailing list discussion).
Seems to work fine for me at least. :)

**Comment by SveSop** [comment]
@AndersDala That is more or less the same code as "staging" uses, but for 3.4 i could not get vsync disabled. The patch [https://github.com/wine-staging/wine-staging/blob/master/patches/wined3d-dxgi_swapchain_Present/0001-wined3d-Implement-updating-swap-interval-through-win.patch](https://github.com/wine-staging/wine-staging/blob/master/patches/wined3d-dxgi_swapchain_Present/0001-wined3d-Implement-updating-swap-interval-through-win.patch) is from staging, and its very similar.

I don't really know why it did not work for me tho, if you have compiled wine-staging-3.4 without doing anything, and test a DX11 (DX9/OpenGL/Vulkan have no problems) benchmark/game and get 60+ fps without troubles.

**Comment by AndersDala** [comment]
I have only tried the patch against wine master, there it works fine for me using DX11 in TW3. 
Tried various build between 3.0 and 3.3, haven't tried on 3.4 as I now get ~80-100FPS using DXVK instead so quite happy. :-)
Haven't used wine-staging lately.

**Comment by SveSop** [comment]
@AndersDala Yeah, i know it works with 3.3 and lower with staging. It have not worked with "release" in DX11 "ever" i think. The staging patches for vsync in dx11 started around 2.16 or so i think.

DXVK uses Vulkan, so is not in any way affected by this.

Are you saying you have had the possibility to disable vsync and get 80-100 fps in DX11 in regular non-staging wine since 3.0? Cos.. if you get 45 fps with "wine-master", then you dont really test the problem :)

**Comment by SveSop** [comment]
Yeah, sunday morning, and i guess missunderstandings is in place of a hefty night out yesterday... 
1. That patch is ALMOST identical to the "staging" patch, that has been included since.. well.. the "new" staging started.. 3.0'ish.
2. That patch (AND the staging one) works on 3.0, 3.1, 3.2 and 3.3, but NO LONGER works on 3.4.
3. Test that patch with 3.4, and tell me if it works or not.

**Comment by AndersDala** [comment]
Haha! 
Sure, I can try it on 3.4 as well. :-)

**Comment by IngeniousDox** [comment]
Very off topic to talk about VSync (with TW3) in "World of Warcraft - Test Result" thread. Has nothing to with WoW or Wine-PBA. Please take it somewhere else.

**Comment by SveSop** [comment]
TestResults = Benchmark = need to disable vsync.
I asked if ppl knew how to get hold of "staging" ppl to fix vsync not being disabled in staging-3.4

It all escalated from there....

**Comment by SveSop** [comment]
And to get back on topic. I tested with wine-staging-3.4+pba +  the hack to disable vsync.
```
Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore) - Min/Max/Avg fps:
28.0/148.5/90.8
Vengeance Point (Broken Shore) -> Crimson Thicket (Suramar) - Min/Max/Avg fps:
30.4/193.4/96.6
```
Slightly less than wine-staging-pba-3.3, but not sure what causes that.

**Comment by SveSop** [comment]
Seems as there was a typo in dlls/wined3d/swapchain.c that did not give "0" as an option to swap_interval setting. 0 being "vsync off". Edited this, and added staging+++, and i could turn on/off vsync in Unigine benchmarks again :)

Posted on a recent bug with a patch, so lets see if its picked up, or if im totally off here hehe.
[https://bugs.winehq.org/show_bug.cgi?id=44623#c6](https://bugs.winehq.org/show_bug.cgi?id=44623#c6)

**Comment by isugimpy** [comment]
@SveSop What hardware are you running on to get numbers like that?  At preset 7, with a stock i5-4690k, 24GB of RAM, a GTX 1080, wine-staging-pba-3.3, nvidia-390.42, I'm getting ~45 FPS average out in the world.  I'll do a proper test of it tonight, and see about the vsync patch as well, but that seems like a drastically different performance profile than I'm seeing.

**Comment by IngeniousDox** [comment]
@isugimpy Quick question, do you have CSMT enabled? Your numbers feel like non csmt/non pba numbers for that hardware.

For the record: I have the same cpu as you, but with 970gtx.

**Comment by FexelFox** [comment]
@IngeniousDox Huh, that certainly cleared up some things for me regarding the intermittent stuttering. I had a hunch it was due to shader caching. Though performance does tank no matter what when using maxed out view distance in stormwind.

**Comment by SveSop** [comment]
@isugimpy Yeah, its no problem to get wow to take a fps nosedive no matter what hardware you have, or if running natively in windows.

Timetest have its own issues with getting stable consistent results, BUT the results are atleast somewhat indicative to performance, and as you can see you have slightly better than me with slightly better hardware :)

Wine performance is very dependant on cpu clockspeed tho, so a stock i5-4690K running at 3.9Ghz(default?) might not speed ahead of my crappy old i7-2600K running at 4.2Ghz

**Comment by IngeniousDox** [comment]
3.4ghz default. I'm running it at 4k Ghz on Air 24/7, could increase it when I increase volume, but I like to stay below 80 degrees on full load.

**Comment by mrdeathjr28** [comment]
@SveSop

Yeah for disgrace wine in most cases depend of higher frecuencies and higher ipc in few cores

For example you can have i5 4690K at 3.9ghz (similar core ryzen performance) in all cores but a i3 8350K (with only 3 cores enabled) at 5.1ghz must be up so much

**Comment by IngeniousDox** [comment]
Staging 3.5 + Firerat Rebase & GEO/CB envvars patch @ GEO 1532, CB 256:

(Latest Liq kernel + Nvidia 390.48 driver)

Timetest Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore) - Min/Max/Avg fps:
```
CSMT: 18 / 121 / 89
CSMT+NV: 17 / 116 / 89
```
And back:
```
CSMT: 18 / 160 / 99
CSMT+NV: 18 / 161 / 96
```
CPU usage is lower, and I don't think I'm hitting 100% gpu either anymore. So I decided to test __GL_THREADED_OPTIMIZATIONS=1 aswell. Still gives me slightly lower fps, so I'm leaving it off for now.

Max VRAM was 1450 MB ish, so I don't think I ever got near the max of heaps, but it didn't hurt to increase the values.

Conclusion: I think I gained a bit of fps compared to the last time I tested. So I advise people to build and test 3.5 for themselves.

**Comment by SveSop** [comment]
Did some testing today, and as usual, not really that conclusive in wow when it came to results, as i could not always get higher result with 1532/256 value... but it DID seem a bit better tbh.

Crimson Thicket (Suramar) -> Vengeance Point (Broken Shore) - Min/Max/Avg fps:
```
31.0/149.4/90.3 - default
33.4/151.5/93.2 - 1532/256
30.4/147.6/93.7 - 256/64
```
Vengeance Point (Broken Shore) -> Crimson Thicket (Suramar) - Min/Max/Avg fps:
```
22.8/188.7/96.1 - default
34.5/193.3/99.0 - 1532/256
30.4/196.7/98.3 - 256/64
```
I also did some tests in Unigine Valley, and there the highest setting of 1532/256 showed a huge drop in fps...
Valley:
```
92.5 fps (3870) 34.6/170.5 - default
75.9 fps (3175) 30.7/113.8 - 1532/512
92.6 fps (3874) 34.2/173.5 - 256/64
```
Slight increase by lowering the value.. Not really sure what this means..

Another knob to tune i guess :)

**Comment by Firerat** [comment]
@SveSop  your not likely to see any real performance gains tweaking the geo heap size, the defaults are probably just fine
to illustrate , try setting geo heap to something very low
say, 5mb
you should see something like this
```
0009:fixme:d3d_perf:buffer_resource_sub_resource_map Failed to allocate new buffer, falling back to sync path.
0030:fixme:d3d_perf:wined3d_buffer_map Fences not used for persistent buffer maps on CS thread, using glFinish (flags: 40002000)
0009:fixme:d3d_perf:wined3d_buffer_heap_alloc Forcing coalesce, not enough free space in buffer heap.
0009:fixme:d3d_perf:wined3d_buffer_heap_deferred_coalesce Performed 0 coalesces.

```
and WoW will stutter 

unless you are seeing msgs like that under normal conditions, upping the heap is not going to help you any

**Comment by SveSop** [comment]
@Firerat Still.. it was quite a performance degradation setting this value "too high" in Unigine Benchmark :)

**Comment by isugimpy** [comment]
Non-objective test in this comment.

I've been playing on the Alpha using these patches, and was getting pretty bad performance, like 30-40FPS in the open world with no other players around.  As I was tweaking things, I was testing by logging the same character in over and over again at the same location, to see if things changed.  That was some of the early testing I mentioned in this issue.

Today, I upgraded to 3.5 with PBA, and created a new wineprefix for it, without CSMT enabled.  Now I'm getting 60+ FPS in the same spot.  I don't know what the hell changed, but it's impressive.  In Dalaran, on live, I was getting about the same 30-40, and now it's hovering in the 55+ range.

I did run a timetest as well, and my average was up about 5 frames from where it was previously.  CPU is about the same, 95%ish, but performance is dramatically improved.

**Comment by isugimpy** [comment]
That actually explains some errors I was getting earlier while I was trying to upgrade to 3.5.  Good to know!

**Comment by SveSop** [comment]
CSMT checkbox is fixed in wine-staging now. [https://github.com/wine-staging/wine-staging/commit/6e3fbe28e1c51d1906624aabbfd3e7724790304c](https://github.com/wine-staging/wine-staging/commit/6e3fbe28e1c51d1906624aabbfd3e7724790304c)
